### PR TITLE
Add support for Quoted Identifiers

### LINKED ISSUE #18720: QuotedIdentifier supports only alphanumeric chars and .
**Description:**
Related PR - https://github.com/ballerina-platform/ballerina-lang/pull/18246. 

JVM Code generations fail for certain chars. Tooling requires additional support from the compiler side in order to render the original identifier. Due to these limitations, it was decided to support only alphanumeric chars and `.` for JBallerina 1.0.0 release.

### PR DESCRIPTION
**Pull request by HindujaB**
Fixes #18720 , #18158 , #19478 , #24951 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anupama-pathirage) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java:2]: Incorrect year?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java [hunk 2]
```diff
@@ -1075,8 +1075,57 @@ private boolean isIdentifierInitialChar(int c) {
...
+            return false;
+        }
+
+        //check unicode private use char
+        if (isUnicodePrivateUseChar(c) || isUnicodePatternWhiteSpaceChar(c)) {
+            return false;
+        }
+
+        //TODO: if (UnicodePatternSyntaxChar) return false
+        return true;
```
Comment (anupama-pathirage) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java:1102]: Can we simplify the above two return statements? 
i.e. `return !isUnicodePrivateUseChar(c) && !isUnicodePatternWhiteSpaceChar(c);`

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    @Test(description = "Test accessing variable in other packages defined with identifier literal")
+    public void testAccessingVarsInOtherPackage() throws BallerinaTestException {
+        Path projectPath = Paths.get(testFileLocation, "TestProject")
+                .toAbsolutePath();
+        String runLog = "Values returned successfully";
+        LogLeecher runLeecher = new LogLeecher(runLog);
+        BMainInstance bMainInstance = new BMainInstance(balServer);
+        bMainInstance.runMain(projectPath.toString(), "pkg.main", new LogLeecher[]{runLeecher});
+        Assert.assertEquals(runLeecher.isTextFound(), true);
```
Comment (anupama-pathirage) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java:173]: Can simplify this to `assertTrue`

File: tests/jballerina-integration-test/src/test/resources/identifier/TestProject/Ballerina.toml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[project]
+org-name = "expression-literal"
+version = "1.0.0"
```
Comment (anupama-pathirage) [inline tests/jballerina-integration-test/src/test/resources/identifier/TestProject/Ballerina.toml:3]: Missing new line at the end.

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java (hunk 1 above)
Comment (HindujaB) [reply tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java]: modified in last commit

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java (hunk 3 above)
Comment (HindujaB) [reply tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/identifier/IdentifierLiteralTestCase.java]: modified in last commit

File: tests/jballerina-integration-test/src/test/resources/identifier/TestProject/Ballerina.toml (hunk 4 above)
Comment (HindujaB) [reply tests/jballerina-integration-test/src/test/resources/identifier/TestProject/Ballerina.toml]: modified in last commit

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java (hunk 2 above)
Comment (HindujaB) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java]: I thought these checks are a bit more clear to compare with the specs in this way. And I referred the implementations of similar logics. [[1]](https://github.com/ballerina-platform/ballerina-lang/blob/cf98e6cf208a5f211e33f0a04f42d22d6118580a/compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java#L1064)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 5]
```diff
@@ -3930,9 +3930,6 @@ private BLangIdentifier createIdentifier(DiagnosticPos pos, String value, Set<Wh
         }
 
         if (value.startsWith(IDENTIFIER_LITERAL_PREFIX)) {
-            if (!escapeQuotedIdentifier(value).matches("^[0-9a-zA-Z.]*$")) {
-                dlog.error(pos, DiagnosticCode.IDENTIFIER_LITERAL_ONLY_SUPPORTS_ALPHANUMERICS);
-            }
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:3935]: Can remove `escapeQuotedIdentifier()` method as well

Comment (HindujaB) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:3935]: removed in the last commit

File: tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    public remote function action1() returns (string) {
+        string 'sample_String_1 = "this ";
+        string '\\\|\ \!\#\$\.\[\;\{\"\:\<\>\u2324_var = "is ";
+        string 'üňĩćőđę_var = "action 1";
+        return 'sample_String_1 + '\\\|\ \!\#\$\.\[\;\{\"\:\<\>\u2324_var + 'üňĩćőđę_var ;
+    }
+
+    public remote function '\u2324_\"\{\ \|\\\!\[\#\$\;_ƮέŞŢ_Action() returns (string){
```
Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal:123]: Add space between function definition and {

Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal:123]: remove brackets for returning one arg

File: tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        string 'sample_String_2 = "this ";
+        string '\\\|\ \!\#\$\.\[\;\{\"\:\<\>\u2324_var2 = "is ";
+        string 'üňĩćőđę_var2 = "action 2";
+        return 'sample_String_2 + '\\\|\ \!\#\$\.\[\;\{\"\:\<\>\u2324_var2 + 'üňĩćőđę_var2 ;
```
Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal:127]: no space before ;

File: tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+function useILAsWorkerName(){
+     worker '\u2324\ 1\!\$\.\[\;\:\<_ƮέŞŢ_Worker returns (string) {
+         string 'var1_\!\$\.\[\;\:\<_ƮέŞŢ = "sample result";
+         return "this is a " + 'var1_\!\$\.\[\;\:\<_ƮέŞŢ;
+     }
+     string result  = wait '\u2324\ 1\!\$\.\[\;\:\<_ƮέŞŢ_Worker;
+     assertEquality("this is a sample result", result);
+}
```
Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_positive.bal:157]: add test for worker interaction

File: tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_service.bal [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+listener http:Listener testEP = new(9090);
+
+@http:ServiceConfig {
```
Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_service.bal:21]: check without giving paths

Comment (HindujaB) [reply tests/jballerina-integration-test/src/test/resources/identifier/identifier_literal_service.bal:21]: User is expected to provide path or encoded name when having reserved characters in the path

**Comment by warunalakshitha** [comment]
Looks to fine for me.

**Comment by lochana-chathura** [comment]
> Parser changes look good.
> 
> However, since quoted-identifier is not an expression, its better to move the parser tests to the miscellaneous category (under `io.ballerinalang.compiler.parser.test.syntax.misc` package).
> May be we can fix that in a separate PR.

@HindujaB  change added with 1fd0392 in PR #25242

**Pull request merged**